src/foam/nanos/auth/PropertyUserPermissionDAO.js

```diff
@@ -0,0 +1,221 @@
+foam.CLASS({
+    package: 'foam.nanos.auth',
+    name: 'PropertyUserPermissionDAO',
+  
+    javaImports: [
+      'foam.core.FObject',
+      'foam.core.PropertyInfo',
+      'foam.core.X',
+      'foam.dao.ProxySink',
+      'foam.nanos.auth.AuthService',
+      'java.util.Iterator',
+      'java.util.LinkedHashMap',
+      'java.util.LinkedList',
+      'java.util.List',
+      'java.util.Map'
+    ],
+  
+    extends: 'foam.dao.ProxyDAO',
+
+    documentation: `A DAO decorator that prevents users from updating / reading
+        properties for which they do not have the update / read permission`,
+
+    methods: [
+      {
+        name: 'put_',
+        javaCode: `
+    FObject oldObj = getDelegate().find(obj.getProperty("id"));
+    return super.put_(x, clearProperties(x, obj, oldObj));
+        `,
+      },
+  
+      {
+        name: 'find_',
+        javaCode: `
+    FObject oldObj = getDelegate().find(id);
+    
+    if ( oldObj != null ) {
+      return hideProperties(x, oldObj);
+    }
+
+    return null;
+        `,
+      },
+
+      {
+        name: 'select_',
+        javaCode: `
+    if ( sink != null ) {
+      ProxySink proxySink = new ProxySink(x, sink) {
```

Comment:
Creating an in-line sub-class is going to prevent this sink from being marshallable over the network, which will prevent this code from working with networked DAO's which will prevent some forms of clustering and scaling. Better to create the Sink as its own model.